### ISSUE TITLE
Support for Kubernetes services

### ISSUE THREAD

**Issue by quickstar** [dapr/dapr#3385]
## In what area(s)?
/area test-and-release

## Describe the feature
As of today the load balancing between Pods is done by dapr and therefore no Kubernetes services are involved. This makes great sense since the K8s svc's are very basic and in most cases they're completely redundant.

Since the development of microservices can be challenging in terms of environmental dependencies, tools like [Telepresence](https://www.telepresence.io) come in very handy.

Unfortunately it seems we cannot use such tools since they're using K8s services to intercept traffic to pods.

I think it would be very practical if we could configure dapr to forward the traffic to k8s services rather then directly to pods. Maybe this would also solve some issues mentioned in https://github.com/dapr/dapr/issues/978

What do you think?

**Comment by liamgib** [comment]
Are you referring to the communication between a sidecar and an application pod? or the communication between services?

Currently, dapr will register the sidecar as a Kubernetes service under "<daprServiceName>-dapr", and the sidecars will use this to invoke other services. I haven't tried telepresence before, but I believe this should already be possible with the existing infrastructure layout.

Does this not work? 🤔

**Comment by yaron2** [comment]
> Are you referring to the communication between a sidecar and an application pod? or the communication between services?
> 
> Currently, dapr will register the sidecar as a Kubernetes service under "-dapr", and the sidecars will use this to invoke other services. I haven't tried telepresence before, but I believe this should already be possible with the existing infrastructure layout.
> 
> Does this not work? 🤔

That is correct.

@quickstar, Dapr indeed does load balancing by keeping a list of endpoints from Kubernetes services for gRPC dns based load balancing, but it still utilizes Kubernetes services under the hood.

**Comment by quickstar** [comment]
> Are you referring to the communication between a sidecar and an application pod? or the communication between services?

I'm referring to the communication in general.
 
> Currently, dapr will register the sidecar as a Kubernetes service under "-dapr", and the sidecars will use this to invoke other services. I haven't tried telepresence before, but I believe this should already be possible with the existing infrastructure layout.
> 
> Does this not work? thinking

I think what this would enable us to do is to intercept the service for the dapr sidecar (I wasn't able to intercept the dapr sidecar). What I try to do is to intercept just my "application-pod" with Telepresence in order to be able to debug only this pod locally.

If there is a way how to do this with Telepresence as of today, I wasn't able to figure it out yet. Any ideas what to try?

**Comment by quickstar** [comment]
According to the [Dapr community call 42](https://www.youtube.com/watch?v=ZwFOEUYe1WA&t=108s),  [Bridge to Kubernetes](https://github.com/microsoft/mindaro) now seems to support K8s. Which is exactly what I was looking for! Very cool, I'll give it a try the next days 👍

**Comment by zivfuture** [comment]
Hi @quickstar , Have you tried  "bridge to kubernetes" ?    It doesn't  work for me.

**Comment by quickstar** [comment]
Hi @zivfuture, Yes it's running but you have to do a proper configuration which can be tricky especially if you're using devcontainers.

# Requirements to debug any service in our kubernetes cluster

> This documentation assumes that the k8s debugging will be run from a devcontainer, therefore the snippets are targetted for those environments.


1. Make sure that kubectl is installed inside the devcontainer
    ```docker
    RUN curl -LO "https://dl.k8s.io/release/$(curl -L -s https://dl.k8s.io/release/stable.txt)/bin/linux/amd64/kubectl"
    RUN install -o root -g root -m 0755 kubectl /usr/local/bin/kubectl
    ```
2. Mount the kubernetes config from the host
    ```yaml
      dev-environment:
        ...
        volumes:
        ...
        - "~/.kube:/home/{USERNAME}/.kube"
    ```

    > It is currently required to use a basic auth login / user since the [Kubelogin](https://github.com/int128/kubelogin) plugin is not yet fully supported.
    
3. Add the [Kubernetes](https://marketplace.visualstudio.com/items?itemName=ms-kubernetes-tools.vscode-kubernetes-tools) and [Bridge to Kubernetes](https://marketplace.visualstudio.com/items?itemName=mindaro.mindaro) extensions to VS Code

    ```json
    "extensions": [
		...
		"ms-kubernetes-tools.vscode-kubernetes-tools",
		"mindaro.mindaro"
	],
    ```

4. Start the devcontainer and configure the required debug configurations

    > The configuration files below are references how they should look like at the end. They can easily be generated via the extensions: (CTRL+P) -> Kubernetes: Debug (Local Tunnel).    This command will guide your through the required steps and generate the configurations as seen below.

    tasks.json
    ```json
    "tasks": [
		...
		{
			"label": "bridge-to-kubernetes.resource",
			"type": "bridge-to-kubernetes.resource",
			"resource": "{SERVICE-TO-DEBUG}",
			"resourceType": "service",
			"ports": [
				8585
			],
			"targetCluster": "staging", // Should not change
			"targetNamespace": "{YOURK8SNAMESPACE}", // Should not change
			"useKubernetesServiceEnvironmentVariables": true // MUST be true for devcontainers!
		}
        ...
	]
    ```

    launch.json
    ```json
    "configurations": [
        ...
        {
            "name": "Launch file with Kubernetes",
            ...
            "preLaunchTask": "bridge-to-kubernetes.resource",
            "env": {
                "GRPC_DNS_RESOLVER": "native"
            }
        }
        ...
    ]
    ```

5. Make sure other services and sidecars are accessed via [Kubernetes service environment variables](https://docs.microsoft.com/en-us/visualstudio/bridge/kubernetes-environment-variables)

If a service must also be available for local debugging (e.g if you want to use the database from the cluster and not the local one) it has to be configured first. See [Configure Bridge to Kubernetes](https://docs.microsoft.com/en-us/visualstudio/bridge/configure-bridge-to-kubernetes)

TL;DR;

Each service dependency we want to use from the cluster and not local must be configured in this configuration. Otherwise the kubernetes service environment variables will not be rewritten correctly!

# Known issues

## Kubelogin not supported
[Kubelogin](https://github.com/int128/kubelogin) is not yet supported, so you have to use a normal user / login for [Bridge to Kubernetes](https://marketplace.visualstudio.com/items?itemName=mindaro.mindaro) to be working.

## Readiness / Lifeness probes via gRPC
While those probes are working correctly during normal deployments, they behave different during debugging. The issues is that [Bridge to Kubernetes](https://marketplace.visualstudio.com/items?itemName=mindaro.mindaro) replaces our deployment with a remote agent, therefor our own docker image which contains the `grpc_health_probe` binary is no longer available. A workaround is to mount this binary via a volume.

```yaml
spec:
  ...
  template:
   ...
    spec:
      ...
      initContainers:
        ...
        - image: alpine:latest
          imagePullPolicy: Always
          name: tagsvc-grpc-health-probe-provider
          command: ["/bin/sh"]
          args: ["-c", "wget -qO/bin/grpc_health_probe https://github.com/grpc-ecosystem/grpc-health-probe/releases/download/v0.4.6/grpc_health_probe-linux-amd64 && chmod +x /bin/grpc_health_probe && cp /bin/grpc_health_probe /grpcHealth/"]
          volumeMounts:
            - name: tagsvc-grpc-health-probe-volume
              mountPath: /grpcHealth
      containers:
        ...
          volumeMounts:
            - name: tagsvc-grpc-health-probe-volume
              mountPath: /usr/local/bin/grpcHealth
              readOnly: true
          ...
          readinessProbe:
            exec:
              command: ["/usr/local/bin/grpcHealth/grpc_health_probe", "-addr=:8585"]
            initialDelaySeconds: 5
          livenessProbe:
            exec:
              command: ["/usr/local/bin/grpcHealth/grpc_health_probe", "-addr=:8585"]
            initialDelaySeconds: 10
      volumes:
        - name: tagsvc-grpc-health-probe-volume
          emptyDir: {}
```

TL;DR

If gRPC health checks are in place, you have to mount the `grpc_health_probe` via volume since the original image will be replaced with the remote agent.
